diff --git a/node_build/make.js b/node_build/make.js
index 522c185..9bd81d5 100644
--- a/node_build/make.js
+++ b/node_build/make.js
@@ -297,6 +297,8 @@ Builder.configure({
                     'BUILDTYPE=Release',
                     'CC=' + builder.config.gcc,
                     'CXX=' + builder.config.gcc,
+                    'LINK=' + builder.config.gcc,
+                    'LDFLAGS=' + builder.config.ldflags,
                     'V=1'
                 ];
 
