diff --git a/setup.py b/setup.py
index 65619dd..a88dc4c 100644
--- a/setup.py
+++ b/setup.py
@@ -5,7 +5,7 @@
 
 
 import os
-from setuptools import setup, find_packages, Command
+from distutils.core import setup, Command
 import sys
 
 from gunicorn import __version__
@@ -76,7 +76,7 @@ setup(
 
     classifiers = CLASSIFIERS,
     zip_safe = False,
-    packages = find_packages(exclude=['examples', 'tests']),
+    packages = ['gunicorn'],
     include_package_data = True,
 
     tests_require = tests_require,
