--- a/node_build/make.js
+++ b/node_build/make.js
@@ -54,7 +54,7 @@ Builder.configure({
         '-std=c99',
         '-Wall',
         '-Wextra',
-        '-Werror',
+        //'-Werror',
         '-Wno-pointer-sign',
         '-pedantic',
         '-D',builder.config.systemName + '=1',
@@ -64,9 +64,9 @@ Builder.configure({
         // Broken GCC patch makes -fstack-protector-all not work
         // workaround is to give -fno-stack-protector first.
         // see: https://bugs.launchpad.net/ubuntu/+source/gcc-4.5/+bug/691722
-        '-fno-stack-protector',
-        '-fstack-protector-all',
-        '-Wstack-protector',
+        //'-fno-stack-protector',
+        //'-fstack-protector-all',
+        //'-Wstack-protector',
 
         '-D','HAS_BUILTIN_CONSTANT_P',
 
@@ -78,13 +78,13 @@ Builder.configure({
         // f8 = 241 peers max, fixed width 8 bit
         // v3x5x8 = 256 peers max, variable width, 3, 5 or 8 bits plus 1 or 2 bits of prefix
         // v4x8 = 256 peers max, variable width, 4, or 8 bits plus 1 bit prefix
-        '-D',' NumberCompress_TYPE=v4x8',
+        '-D',' NumberCompress_TYPE=v4x8'
 
         // disable for speed, enable for safety
-        '-D','Log_DEBUG',
-        '-D','Identity_CHECK=1',
-        '-D','Allocator_USE_CANARIES=1',
-        '-D','PARANOIA=1'
+        //'-D','Log_DEBUG',
+        //'-D','Identity_CHECK=1',
+        //'-D','Allocator_USE_CANARIES=1',
+        //'-D','PARANOIA=1'
     );
     if (process.env['NO_PIE'] === undefined) {
         builder.config.cflags.push('-fPIE');
